File: lintcode/cpp/First_Missing_Positive.cpp
class Solution {
public:
    /**    
     * @param A: a vector of integers
     * @return: an integer
     */
    int firstMissingPositive(vector<int> A) {
        if(A.empty()) return 1;
        int n = A.size();
        
        for(int i = 0; i < n; i++){
            if(A[i] <= 0) A[i] = n+1;
        }
        
        for(int i = 0; i < n; i++){
            int k = abs(A[i]);
            if(k <= n && A[k-1] > 0) A[k-1] *=-1;
        }
        
        for(int i = 0; i < n; i++){
            if(A[i] > 0) return i+1;
        }
        return n+1;
    }
};

File: lintcode/java/FirstMissingPositive.java
public class Solution {
    /**    
     * @param A: an array of integers
     * @return: an integer
     */
    public int firstMissingPositive(int[] A) {
        Arrays.sort(A);
        
        int ind = 1;
        for(int x : A){
            if(x <= 0) continue;
            else {
                if(x == ind) ind++;
                else if(x < ind) continue;
                else return ind;
            }
        }
        return ind;
    }
}
